File: src/main/java/nnglebanov/daoexample/rest/dto/BookDto.java
package nnglebanov.daoexample.rest.dto;

import lombok.Data;
import nnglebanov.daoexample.domain.Book;

import java.util.List;
import java.util.stream.Collectors;

@Data
public class BookDto {
    private Integer id;
    private String bookTitle;
    private List<AuthorDto> authors;


    public static BookDto toDto(Book book) {
        BookDto bookDto = new BookDto();
        bookDto.setId(book.getId());
        bookDto.setBookTitle(book.getBookTitle());
        bookDto.setAuthors(
                book.getAuthors().stream()
                        .map(AuthorDto::toDto)
                        .collect(Collectors.toList())
        );
        return bookDto;
    }

    public static Book toEntity(BookDto bookDto) {
        Book book = new Book();
        book.setBookTitle(bookDto.getBookTitle());
        book.setAuthors(bookDto.getAuthors().stream().map(AuthorDto::toEntity).collect(Collectors.toList()));
        return book;
    }
}

File: src/main/java/nnglebanov/daoexample/rest/dto/BookDto.kt
package nnglebanov.daoexample.rest.dto

import nnglebanov.daoexample.domain.Book
import java.util.*

class BookDto(var id: String? = null,
              var bookTitle: String? = null,
              var authors: List<AuthorDto>? = null) {


    companion object {


        fun toDto(book: Book): BookDto {
            return BookDto().apply {
                id = book.id
                bookTitle = book.bookTitle
                authors = book.authors?.map { author -> AuthorDto.toDto(author) }?.toList() ?: listOf()
            }
        }

        fun toEntity(bookDto: BookDto): Book {
            return Book(
                    bookTitle = bookDto.bookTitle,
                    authors = bookDto.authors?.map { authorDto -> AuthorDto.toEntity(authorDto) }?.toMutableList(),
                    genres = mutableListOf(),
                    comments = mutableListOf(),
                    dateTime = Date()
            )

        }
    }
}